sdk/go/pulumi/context.go

```diff
@@ -177,6 +177,12 @@ func (ctx *Context) Context() context.Context {
 	return ctx.ctx
 }
 
+// Returns a copy of base context in which the value associated with key is val.
+func (ctx *Context) WithValue(key, val any) context.Context {
+	ctx.ctx = context.WithValue(ctx.ctx, key, val)
```

Comment:
This mutates the current pulumi context. When we've discussed this internally the expecation was that `pulumi.Context.WithValue` would return a new `pulumi.Context` but such that if you then called `Context()` on that you'd get a `context.Context` with the value set. i.e. The same semantics as the normal `context.Context.WithValue` but wrapped by the `pulumi.Context`.

Turns out doing all that correctly is non-trivial, thus why this feature hasn't been picked up for so long.

Reply:
_non-trivial_, so do you feel like this is feature worth putting time into since it is hard and small? I haven't seen much traction on it. Since [the task](https://github.com/pulumi/pulumi/issues/11825) it came from had a label `good-first-issue` I figured it was that simple but if not than maybe I could pick something else :)

Reply:
> so do you feel like this is feature worth putting time into since it is hard and small?

It would be nice, and it's probably not _that_ hard. Just not trivial.

>  I figured it was that simple

It might be. I think it might be doable by splitting `Context` in two. One field to a `context.Context` that `WithValue` could change and copy to a new `pulumi.Context`, and one filed pointing to all the other state currently in Context which would be shared by reference by all Context's created by WithValue.

Not trivial because that's a fair bit of code to fix up to deal with that, but it is fairly simple, just split the type in two.

Reply:
I think I see. I'll try to do this then. Any idea how to do tests for it? or just manual ones?

Reply:
Can probably add something to sdk/go/pulumi/context_test.go to test it.